// ts-css-modules-vite-plugin/dist/vite.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_react_1 = __importDefault(require("@vitejs/plugin-react"));
const vite_plugin_svgr_1 = __importDefault(require("vite-plugin-svgr"));
const vite_1 = require("vite");
const vite_plugin_sass_dts_1 = __importDefault(require("vite-plugin-sass-dts"));
const vite_plugin_dts_1 = __importDefault(require("vite-plugin-dts"));
const vite_tsconfig_paths_1 = __importDefault(require("vite-tsconfig-paths"));
const pkg = require("./package.json");
// @ts-ignore
const path_1 = __importDefault(require("path"));
exports.default = (0, vite_1.defineConfig)({
    base: './',
    plugins: [
        (0, plugin_react_1.default)(),
        (0, vite_plugin_svgr_1.default)({
            svgrOptions: {
                exportType: 'named', jsxRuntime: 'automatic'
            }
        }),
        (0, vite_tsconfig_paths_1.default)(),
        (0, vite_plugin_sass_dts_1.default)({
            global: {
                generate: true,
                outFile: path_1.default.resolve("/Users/richardmiles/IdeaProjects/node-core/node_modules/@drop-in-gaming/core", "./src/style.d.ts"),
            },
        }),
        (0, vite_plugin_dts_1.default)({
            insertTypesEntry: true
        }),
    ],
    resolve: {
        alias: [
            {
                find: '@',
                replacement: path_1.default.resolve(__dirname, 'src')
            },
        ]
    },
    css: {
        preprocessorOptions: {
            scss: {
                importer(...args) {
                    return {
                        file: `${path_1.default.resolve(__dirname, './src')}`
                    };
                }
            }
        },
        modules: {
            localsConvention: 'camelCase'
        }
    },
    build: {
        assetsDir: 'src/assets',
        sourcemap: true,
        minify: false,
        lib: {
            entry: path_1.default.resolve(__dirname, 'src/index.ts'),
            name: '@drop-in-gaming/core',
            formats: ['es', 'umd', 'cjs'],
            fileName: (format) => format === 'es' ? `core.${format}.jsx` : `core.${format}.js`,
        },
        rollupOptions: {
            external: [
                ...Object.keys(pkg.peerDependencies)
            ],
            output: {
                globals: {
                    react: 'React',
                    'react-dom': 'ReactDOM',
                    'styled-components': 'styled',
                    axios: 'axios',
                    'axios-cache-adapter': 'axios-cache-adapter',
                },
                minifyInternalExports: false
            },
        },
    }
});
